peer/connector.go

```diff
@@ -80,48 +83,91 @@ func (conn *connector) initialise(privateKey []byte, publicKey []byte, connect [
 	}
 	conn.staticClients = make([]*upstream.Upstream, staticCount)
 
-	// error code for goto fail
-	errX := error(nil)
-
 	// initially connect all static sockets
+	wg := sync.WaitGroup{}
+	errCh := make(chan error, len(connect))
+
+	conn.log.Debugf("static connection count: %d", len(connect))
+
 	for i, c := range connect {
-		address, err := util.NewConnection(c.Address)
-		if nil != err {
-			log.Errorf("client[%d]=address: %q  error: %s", i, c.Address, err)
-			errX = err
-			goto fail
-		}
-		serverPublicKey, err := hex.DecodeString(c.PublicKey)
-		if nil != err {
-			log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
-			errX = err
-			goto fail
-		}
+		wg.Add(1)
 
-		// prevent connection to self
-		if bytes.Equal(publicKey, serverPublicKey) {
-			errX = fault.ErrConnectingToSelfForbidden
-			log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, errX)
-			goto fail
-		}
+		// start new goroutine for each connection
+		go func(conn *connector, c Connection, i int, wg *sync.WaitGroup, ch chan error) {
 
-		client, err := upstream.New(privateKey, publicKey, connectorTimeout)
-		if nil != err {
-			log.Errorf("client[%d]=%q  error: %s", i, address, err)
-			errX = err
-			goto fail
-		}
+			// error function call
+			errF := func(wg *sync.WaitGroup, ch chan error, e error) {
+				ch <- e
+				wg.Done()
+			}
 
-		conn.staticClients[i] = client
-		globalData.connectorClients = append(globalData.connectorClients, client)
+			// for canonicaling the error
+			canonicalErrF := func(c Connection, e error) error {
+				return fault.GenericError(fmt.Sprintf("client: %q error: %s", c.Address, e))
+			}
 
-		err = client.Connect(address, serverPublicKey)
-		if nil != err {
-			log.Errorf("connect[%d]=%q  error: %s", i, address, err)
-			errX = err
-			goto fail
-		}
-		log.Infof("public key: %x  at: %q", serverPublicKey, c.Address)
+			address, err := util.NewConnection(c.Address)
+			if nil != err {
+				log.Errorf("client[%d]=address: %q  error: %s", i, c.Address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+			serverPublicKey, err := hex.DecodeString(c.PublicKey)
+			if nil != err {
+				log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			// prevent connection to self
+			if bytes.Equal(publicKey, serverPublicKey) {
+				err := fault.ErrConnectingToSelfForbidden
+				log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			client, err := upstream.New(privateKey, publicKey, connectorTimeout)
+			if nil != err {
+				log.Errorf("client[%d]=%q  error: %s", i, address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			conn.Lock()
+			conn.staticClients[i] = client
+			conn.Unlock()
+
+			globalData.connectorClients = append(globalData.connectorClients, client)
```

Comment:
Isn't here has a concurrent issue? All go rountines access globalData

Reply:
Yes. I think so. It's my mistake